macaroons/constraints.go

```diff
@@ -0,0 +1,86 @@
+package macaroons
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"gopkg.in/macaroon-bakery.v1/bakery/checkers"
+	macaroon "gopkg.in/macaroon.v1"
+)
+
+// Constraint type adds a layer of indirection over macaroon caveats and
+// checkers.
+type Constraint func(*macaroon.Macaroon)
+
+// AddConstraints returns new derived macaroon by applying every passed
+// constraint and tightening its restrictions.
+func AddConstraints(mac *macaroon.Macaroon, cs ...Constraint) *macaroon.Macaroon {
+	newMac := mac.Clone()
+	for _, constraint := range cs {
+		constraint(newMac)
```

Comment:
Check for error from the `constraint(newMac)` call, and return `nil, err` if `err` isn't nil.